providers/component-protocol/cptype/sdk.go

```diff
@@ -37,19 +38,42 @@ type SDK struct {
 	Event       ComponentEvent
 	CompOpFuncs map[OperationKey]OperationFunc
 	Comp        *Component
+
+	// for parallel use, it's request level
+	StdStructuredPtr IStdStructuredPtr
+
+	Lock       sync.Mutex
+	OriginLock *sync.Mutex
 }
 
 // Clone only return general-part of sdk to avoid concurrency issue.
 func (sdk *SDK) Clone() *SDK {
-	return &SDK{
-		Ctx:         sdk.Ctx,
-		Scenario:    sdk.Scenario,
-		Tran:        sdk.Tran,
-		Identity:    sdk.Identity,
-		InParams:    sdk.InParams,
-		Lang:        sdk.Lang,
-		GlobalState: sdk.GlobalState,
+	sdk.Lock.Lock()
+	defer sdk.Lock.Unlock()
+
+	clonedSDK := SDK{
+		Ctx:      sdk.Ctx,
+		Scenario: sdk.Scenario,
+		Tran:     sdk.Tran,
+		Identity: sdk.Identity,
+		Lang:     sdk.Lang,
+
+		OriginLock: &sdk.Lock,
 	}
+	// inParams
+	clonedInParams := make(InParams)
+	for k, v := range sdk.InParams {
+		clonedInParams[k] = v
+	}
+	clonedSDK.InParams = clonedInParams
+	// globalStates
+	clonedGlobalStates := make(GlobalStateData)
+	for k, v := range *sdk.GlobalState {
+		clonedGlobalStates[k] = v
```

Comment:
what if `v` is a pointer?

Reply:
No problem here.
It's a range pitfall, if you don't save address of `v`, it's ok.